cpp/src/genetic/reg_stack.cuh

```diff
@@ -58,38 +58,20 @@ struct stack {
    *       to push more than `MaxSize` elements leads to all sorts of incorrect
    *       behavior.
    */
-  HDI void push(DataT val) {
-#pragma unroll
-    for (int i = 0; i < MaxSize; ++i) {
-      if (elements_ == i) {
-        regs_[i] = val;
-        ++elements_;
-      }
-    }
-  }
+  HDI void push(DataT val) { if(!full()) regs_[++elements_] = val; }
```

Comment:
```suggestion
  HDI void push(DataT val) { 
  #pragma unroll
    for (int i = MaxSize-1; i >= 0; --i) {
      if (elements_ == i) {
        regs_[i] = val;
        ++elements_;
      }
    }
  }
```

Reply:
I'll do this - my own experiments also show `nvcc` (on cuda 11.0) unrolls the loops completely (I did this by viewing the generated ptx code on a separate program which uses `reg_stack.cuh`).